### PR TITLE
[WIP] jdbc operators (pg>, redshift> ) enable logging 

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
Follow-up #629 

This PR introduce logging option to JDBC operators(`pg>` and `redshift>`).

@yoyama Please take a look when you have time.
`AbstractJdbcConnection` need to `debug` option for this feature.
I'm worrying how to enable the debug option. 

This option output logs like the below.

```
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1): pg>: select.sql
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1): SET TRANSACTION READ ONLY
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1): select *
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1):  from users
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1): -- where id = 1
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=0=1): SET TRANSACTION READ WRITE
2019-02-27 20:34:55 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2): pg>: select.sql
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2): SET TRANSACTION READ ONLY
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2): select *
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2):  from users
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2): -- where id = 2
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=1=2): SET TRANSACTION READ WRITE
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3): pg>: select.sql
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3): SET TRANSACTION READ ONLY
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3): select *
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3):  from users
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3): -- where id = 3
2019-02-27 20:34:56 +0900 [INFO] (0017@[0:default]+test+task^sub+for-0=id=2=3): SET TRANSACTION READ WRITE
```

### CODE REVIEW

File: digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/AbstractJdbcJobOperator.java [hunk 1]
```diff
@@ -102,6 +104,7 @@ protected TaskResult run(Config params, Config state, C connectionConfig)
         }
 
         try (JdbcConnection connection = connect(connectionConfig)) {
+            connection.setDebug(isDebug);
```
Comment (hiroyuki-sato) [inline digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/AbstractJdbcJobOperator.java]: I'm worried backward compatibility. so I didn't change constructor.

File: digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java [hunk 2]
```diff
@@ -33,5 +33,8 @@ default String escapeTableReference(TableReference ref)
 
     String escapeIdent(String ident);
 
+    void setDebug(Boolean debug);
+    Boolean getDebug();
+
```
Comment (hiroyuki-sato) [inline digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java]: This change may break compatibility. For ex. Does `digdag-plugin-mysql` work new version?

File: digdag-standards/src/main/java/io/digdag/standards/operator/pg/PgConnection.java [hunk 3]
```diff
@@ -41,6 +41,7 @@ public void executeReadOnlyQuery(String sql, Consumer<JdbcResultSet> resultHandl
         try {
             execute("SET TRANSACTION READ ONLY");
             try (Statement stmt = connection.createStatement()) {
+                loggingExecuteSQL(sql);
                 ResultSet rs = stmt.executeQuery(sql);  // executeQuery throws exception if given query includes multiple statements
```
Comment (hiroyuki-sato) [inline digdag-standards/src/main/java/io/digdag/standards/operator/pg/PgConnection.java]: Almost logging added in `AbstractJdbcConnection.java`. but this method use `stmt.executeQuery` at this place.
That's why I called `loggingExecuteSQL` method in this part.

File: digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java (hunk 2 above)
Comment (yoyama) [reply digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java]: I am also afraid of breaking compatibility.
How about define as default methods as follows.
```
    default void setDebug(boolean debug) { throw new RuntimeException("..."); }

    default boolean getDebug() { return false; }
```

I don't know we can avoid breaking compatibility by default method.
Could you check the binary with digdag-plugin-mysql works or not?

**Comment by yoyama** [comment]
@hiroyuki-sato Thank you for your proposal.
Just idea, how about changing the parameter name from _debug_ to another one. For example _show_query_ or _show_sql_. debug may be ambiguous. But if you are going to add another debug related info with this parameter, debug may be good.
I also comment to code, please check it.

**Comment by hiroyuki-sato** [comment]
@yoyama Thank you for the comment.
I referenced this [comment](https://github.com/treasure-data/digdag/pull/629#issuecomment-335574146). And I used the option name `debug`. 
But I think `show_query`  or `show_sql` is better for this operator too.

**Comment by hiroyuki-sato** [comment]
@yoyama. Please take a look again when you have time.

* Change option name `debug` to `show_query`.
* Define `default` method.
* Add redshift logging.

I'll check backward compatibility later using `digdag-plugin-mysql`.

And one more thing. I don't have a Redshift environment. Can you test for that?

File: digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java (hunk 2 above)
Comment (hiroyuki-sato) [reply digdag-standards/src/main/java/io/digdag/standards/operator/jdbc/JdbcConnection.java]: defined default method.

**Comment by hiroyuki-sato** [comment]
@yoyama 

It seems that this change keeps backward compatibility.

I succeed to run `digdag-plugin-mysql` with this branch.

```yaml
_export:
  plugin:
    repositories:
      #- file://${repos}
      #- file:///path/to/digdag-plugin-mysql/build/repo
      - https://jitpack.io
    dependencies:
      - com.github.hiroyuki-sato:digdag-plugin-mysql:0.1.1

  mysql:
    host: localhost
    user: root
    database: digdag_test
    ssl: true
    show_query: true

+step1:
  mysql>: test.sql
  download_file: test.txt
```

```
./digdag-0.9.35-SNAPSHOT.jar run -a plugin.dig
2019-03-01 21:21:15 +0900: Digdag v0.9.34
2019-03-01 21:21:16 +0900 [WARN] (main): Reusing the last session time 2018-04-25T00:00:00+00:00.
2019-03-01 21:21:16 +0900 [INFO] (main): Using session /path/to/digdag-plugin-mysql/sample/.digdag/status/20180425T000000+0000.
2019-03-01 21:21:16 +0900 [INFO] (main): Starting a new session project id=1 workflow name=plugin session_time=2018-04-25T00:00:00+00:00
2019-03-01 21:21:17 +0900 [INFO] (0017@[0:default]+plugin+step1): mysql>: test.sql
2019-03-01 21:21:18 +0900 [INFO] (0017@[0:default]+plugin+step1): SET TRANSACTION READ ONLY
2019-03-01 21:21:18 +0900 [INFO] (0017@[0:default]+plugin+step1): SET TRANSACTION READ WRITE
Success. Task state is saved at /path/to/digdag-plugin-mysql/sample/.digdag/status/20180425T000000+0000 directory.
  * Use --session <daily | hourly | "yyyy-MM-dd[ HH:mm:ss]"> to not reuse the last session time.
  * Use --rerun, --start +NAME, or --goal +NAME argument to rerun skipped tasks.
```

**Comment by yoyama** [comment]
@hiroyuki-sato Thank you for your modification and confirmed the compatibility with digdag-plugin-mysql. I think it is basically good. 
But I am concerned about _PgConnection.java_ and _RedshiftConnection.java._ Embedding _loggingExecuteSQL()_ to each connection class is complicated.
So another way may be better. For example, Wrap _ResultSet executeQuery(...)_ is one of the way.
In addition it is better if you can add tests for this PR.

**Comment by hiroyuki-sato** [comment]
@yoyama Thanks. what do you think b5c122c change?
I'll add test later.

**Comment by hiroyuki-sato** [comment]
@yoyama Sorry 8aeebb3 renamed that method to `executeQuery()`. And I used try-with-resouces statement in `executeQuery()` method.

**Comment by yoyama** [comment]
I think basically ok but could you update the PR to review?

**Comment by hiroyuki-sato** [comment]
@yoyama Thank you for the comment. I'll modify code style and add tests.

**Comment by hiroyuki-sato** [comment]
@yoyama Could you advise me?

`ResultSet rs = executeQuery(sql);` doesn't work as I expected.
It needs to write inside a try-with-resources block.

Any idea?

```java
protected ResultSet executeQuery(String sql)
          throws SQLException
  {
      try (Statement stmt = connection.createStatement()) {
          loggingExecuteSQL(sql);
          return stmt.executeQuery(sql); // executeQuery throws exception if given query includes multiple statements
      }
  }
	
```

```java
String sql = String.format(ENGLISH,
          "SELECT query_id FROM %s WHERE completed_at < SYSDATE - INTERVAL '%d SECOND'",
          escapeTableReference(statusTable),
          cleanupDuration.getSeconds());
  ResultSet rs = executeQuery(sql); // this part raise an error "This ResultSet is closed."
  if (rs.next()) {
      sql = String.format("DROP TABLE %s", escapeTableReference(statusTable));
      executeUpdate(sql);
  }
```

**Comment by yoyama** [comment]
I will close this ticket because there is no progress for long time.
Feel free to re-open in the future.

**Pull request closed without merging**